### PR TITLE
Production db backup workflow

### PR DESCRIPTION
**Pull request by VenkataChalla554**
## Context

Workflow for production database backup.

## Changes proposed in this pull request

Created New Workflow and scheduled early morning 3 AM to take DB backups to azure storage.

## Guidance to review

 Test is done on this feature branch  : https://github.com/DFE-Digital/itt-mentor-services/actions/runs/8874816721/job/24362956458

## Link to Trello card

https://trello.com/c/QXV3IPgZ/1743-backup-ittms-production-database

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated or [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

**Comment by VenkataChalla554** [comment]
[Backup ITTMS Production Database](https://trello.com/c/QXV3IPgZ/1743-backup-ittms-production-database)

File: .github/workflows/database-backup.yml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+concurrency: build_and_deploy_main
+
+on:
+  workflow_dispatch:
+    inputs:
+      overwriteThisMorningsBackup:
+        required: true
+        type: boolean
+        default: false
+      restoreToProductionDataEnv:
```
Comment (RMcVelia) [inline .github/workflows/database-backup.yml:11]: input not required

File: .github/workflows/database-backup.yml [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        required: true
+        type: boolean
+        default: false
+      restoreToProductionAnalysisEnv:
```
Comment (RMcVelia) [inline .github/workflows/database-backup.yml:15]: input not required

File: .github/workflows/database-backup.yml [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        required: true
+        type: boolean
+        default: false
+      restoreToStagingEnv:
```
Comment (RMcVelia) [inline .github/workflows/database-backup.yml:19]: input not required

File: .github/workflows/database-backup.yml [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        required: true
+        type: boolean
+        default: false
+  schedule: # 03:00 UTC
+    - cron: '0 3 * * *'
+
+jobs:
+  backup:
+    name: Backup AKS Database (production)
+    # if: ${{ github.event_name == 'schedule' || (github.event_name == 'workflow_dispatch' && github.event.inputs.overwriteThisMorningsBackup == 'true') }}
```
Comment (RMcVelia) [inline .github/workflows/database-backup.yml:29]: should be uncommented

File: .github/workflows/database-backup.yml [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    - uses: azure/login@v2
+      with:
+        creds: ${{ secrets.AZURE_CREDENTIALS }}
+
+    - name: Setup postgres client
+      uses: DFE-Digital/github-actions/install-postgres-client@master
+      with:
+        version: 14
+
+
```
Comment (RMcVelia) [inline .github/workflows/database-backup.yml:63]: can you remove the multiple blank lines

File: .github/workflows/database-backup.yml [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+      with:
+        azure-credentials: ${{ secrets.AZURE_CREDENTIALS }}
+
+    - name: K8 setup
+      shell: bash
+      run: |
+        make ci production get-cluster-credentials
+        make install-konduit
+
+    - name: Setup postgres client
```
Comment (RMcVelia) [inline .github/workflows/database-backup.yml:81]: already installed? see line 58

File: .github/workflows/database-backup.yml [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    - name: Backup Prod DB
+      run: |
+        bin/konduit.sh -t 7200 itt-mentor-services-production -- pg_dump -E utf8 --clean --if-exists --no-owner --verbose --no-password -f ${BACKUP_FILE_NAME}.sql
+        tar -cvzf ${BACKUP_FILE_NAME}.tar.gz ${BACKUP_FILE_NAME}.sql
+
+    - name: Set up environment variables
+      shell: bash
+      run: |
+        echo "STORAGE_ACCOUNT_RG=s189p01-ittms-pd-rg" >> $GITHUB_ENV
+        SERVICE_SHORT=ittms
```
Comment (RMcVelia) [inline .github/workflows/database-backup.yml:96]: is SERVICE_SHORT required?

File: Makefile [hunk 8]
```diff
@@ -172,3 +172,9 @@ get-cluster-credentials: set-azure-account
 bin/konduit.sh:
 	curl -s https://raw.githubusercontent.com/DFE-Digital/teacher-services-cloud/main/scripts/konduit.sh -o bin/konduit.sh \
 		&& chmod +x bin/konduit.sh
+.PHONY: install-konduit
```
Comment (RMcVelia) [inline Makefile:175]: this is already in the Makefile as bin/konduit.sh at line 172

File: .github/workflows/database-backup.yml [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+    - uses: DFE-Digital/github-actions/set-kubelogin-environment@master
+      with:
+        azure-credentials: ${{ secrets.AZURE_CREDENTIALS }}
+
+    - name: K8 setup
+      shell: bash
+      run: |
+        make ci production get-cluster-credentials
+        make install-konduit
```
Comment (RMcVelia) [inline .github/workflows/database-backup.yml:79]: should be make bin/konduit.sh

File: .github/workflows/database-backup.yml (hunk 1 above)
Comment (VenkataChalla554) [reply .github/workflows/database-backup.yml]: Removed input

File: .github/workflows/database-backup.yml (hunk 3 above)
Comment (VenkataChalla554) [reply .github/workflows/database-backup.yml]: Removed input

File: .github/workflows/database-backup.yml (hunk 4 above)
Comment (VenkataChalla554) [reply .github/workflows/database-backup.yml]: fixed

File: .github/workflows/database-backup.yml (hunk 6 above)
Comment (VenkataChalla554) [reply .github/workflows/database-backup.yml]: removed duplicate codes

File: Makefile (hunk 8 above)
Comment (VenkataChalla554) [reply Makefile]: used existing one from make file

File: .github/workflows/database-backup.yml (hunk 9 above)
Comment (VenkataChalla554) [reply .github/workflows/database-backup.yml]: ya, removed

File: .github/workflows/database-backup.yml (hunk 7 above)
Comment (VenkataChalla554) [reply .github/workflows/database-backup.yml]: removed this unused

File: .github/workflows/database-backup.yml [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    services:
+      postgres:
+        image: postgres:11.10
+        env:
+          POSTGRES_USER: postgres
+          POSTGRES_PASSWORD: postgres
+          POSTGRES_DB: postgres
+        ports:
+        - 5432:5432
+        options: --health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 5
```
Comment (RMcVelia) [inline .github/workflows/database-backup.yml:25]: I don't believe you need the postgres service, as you are installing the client and we aren't doing any local db work (like sanitise)

File: .github/workflows/database-backup.yml [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    steps:
+    - uses: actions/checkout@v4
+      name: Checkout
+
+    - name: Set KV environment variables
+      run: |
+        tf_vars_file=terraform/application/config/production.tfvars.json
+        echo "namespace=$(jq -r '.namespace' ${tf_vars_file})" >> $GITHUB_ENV
+        echo "cluster=$(jq -r '.cluster' ${tf_vars_file})" >> $GITHUB_ENV
+        echo "app_environment=$(jq -r '.cluster' ${tf_vars_file})" >> $GITHUB_ENV
```
Comment (RMcVelia) [inline .github/workflows/database-backup.yml:35]: I don't think this step (set kv env vars) is required, as I don't see these vars being used anywhere

File: .github/workflows/database-backup.yml [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+on:
+  workflow_dispatch:
+  schedule: # 03:00 UTC
+    - cron: '0 3 * * *'
+
+jobs:
+  backup:
+    name: Backup AKS Database (production)
+    if: ${{ github.event_name == 'schedule' || (github.event_name == 'workflow_dispatch' && github.event.inputs.overwriteThisMorningsBackup == 'true') }}
```
[the remark this answers is not in the document]
Comment (VenkataChalla554) [reply .github/workflows/database-backup.yml:17]: Yes, Added  input overwriteThisMorningsBackup

File: .github/workflows/database-backup.yml (hunk 10 above)
Comment (VenkataChalla554) [reply .github/workflows/database-backup.yml]: removed this part

**Review by RMcVelia — APPROVED** [review]

**Pull request merged**